### ISSUE TITLE
Got Postgrex.Error prepared statement "ecto_insert_posts" does not exist in ecto_sql tests

### ISSUE THREAD

**Issue by Eiji7** [elixir-ecto/ecto_sql#160]
### Environment

I have up to date `openSUSE Tumbleweed`. However I have installed `Erlang`, `Elixir` and `PostgreSQL` using [asdf](https://github.com/asdf-vm/asdf) (extendable version manager).

```bash
$ asdf current
elixir         1.9.2-otp-22 (set by …)
erlang         22.1.5   (set by …)
postgres       12.0     (set by …)
$ elixir -v
Erlang/OTP 22 [erts-10.5.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Elixir 1.9.2 (compiled with Erlang/OTP 22)
```

Steps to reproduce:
1. Install `asdf` with 3 plugins: `erlang`, `elixir` and `postgres`.
2. Install latest versions (as above) from all of those plugins
3. Ensure you have no other instances of `PostgreSQL` running
4. Run `pg_ctl start`
5. Create a `UTF-8` database replacing `username` with name of your `Linux` account (`asdf` prepares account for easy `dev` work)
6. Clone `ecto_sql`
7. Run few times tests

Here is an `SQL` I have used in order to create database:
```sql
create database "username" with owner "username" encoding 'UTF8' LC_COLLATE = 'pl_PL.UTF-8' LC_CTYPE = 'pl_PL.UTF-8' TEMPLATE template0;
```
> Based on: https://support.helpspot.com/index.php?pg=kb.page&id=467

Here is my test command:
```bash
PG_URL="username:@127.0.0.1" ECTO_ADAPTER=pg mix test
```

All of dependencies are up to date as today I was looking for this error on my fork of `master` branch.

Here is a complete list of all dependencies:
```bash
* connection 1.0.4 (Hex package) (mix)
  locked at 1.0.4 (connection) a1cae722
  ok
* telemetry 0.4.0 (Hex package) (rebar3)
  locked at 0.4.0 (telemetry) 8339bee3
  ok
* decimal 1.8.0 (Hex package) (mix)
  locked at 1.8.0 (decimal) ca462e0d
  ok
* db_connection 2.1.1 (Hex package) (mix)
  locked at 2.1.1 (db_connection) a51e8a2e
  ok
* ecto 3.3.0-dev (https://github.com/elixir-ecto/ecto.git) (mix)
  locked at 74cd9f5
  ok
* postgrex 0.15.0 (Hex package) (mix)
  locked at 0.15.0 (postgrex) dd534916
  ok
* myxql 0.2.4 (Hex package) (mix)
  locked at 0.2.4 (myxql) 010c3a14
  ok
```

### Current behavior

Here is the error which happen locally and on CI:

```elixir
  1) test insert autogenerates for custom type (Ecto.Integration.RepoTest)
     deps/ecto/integration_test/cases/repo.exs:226
     ** (Postgrex.Error) ERROR 26000 (invalid_sql_statement_name) prepared statement "ecto_insert_posts" does not exist
     code: post = TestRepo.insert!(%Post{uuid: nil})
     stacktrace:
       (ecto_sql) lib/ecto/adapters/sql.ex:631: Ecto.Adapters.SQL.raise_sql_call_error/1
       (ecto) lib/ecto/repo/schema.ex:651: Ecto.Repo.Schema.apply/4
       (ecto) lib/ecto/repo/schema.ex:262: anonymous fn/15 in Ecto.Repo.Schema.do_insert/4
       (ecto) lib/ecto/repo/schema.ex:164: Ecto.Repo.Schema.insert!/4
       deps/ecto/integration_test/cases/repo.exs:227: (test)
```

possibly related:

```elixir
  1) test insert and update with changeset (Ecto.Integration.RepoTest)
     deps/ecto/integration_test/cases/repo.exs:166
     ** (Postgrex.Error) ERROR 26000 (invalid_sql_statement_name) prepared statement "ecto_update_posts" does not exist
     code: assert %Post{text: "y", title: "world", temp: "unknown"} = TestRepo.update!(changeset)
     stacktrace:
       (ecto_sql) lib/ecto/adapters/sql.ex:629: Ecto.Adapters.SQL.raise_sql_call_error/1
       (ecto) lib/ecto/repo/schema.ex:651: Ecto.Repo.Schema.apply/4
       (ecto) lib/ecto/repo/schema.ex:349: anonymous fn/15 in Ecto.Repo.Schema.do_update/4
       (ecto) lib/ecto/repo/schema.ex:177: Ecto.Repo.Schema.update!/4
       deps/ecto/integration_test/cases/repo.exs:179: (test)
```

As I'm able to reproduce I have inspected the error struct:

```elixir
%Postgrex.Error{
  connection_id: 10209,
  message: nil,
  postgres: %{
    code: :invalid_sql_statement_name,
    file: "prepare.c",
    line: "502",
    message: "prepared statement \"ecto_insert_posts\" does not exist",
    pg_code: "26000",
    routine: "FetchPreparedStatement",
    severity: "ERROR",
    unknown: "ERROR"
  },
  query: nil
}
```

I'm not sure if that would help, but some tests are failing if developer would not create a database, but even that thing is not stopping this error from raising, so maybe it could somehow help by elimination.

### Expected behavior

All tests should pass.

Hope that it helps. I will still look for possible solution, but it may take some time for me as I have not yet contributing to this part of code in this repository. If you are still unable to reproduce then maybe I could help and send some inspected results that you may find interesting.

**Comment by Eiji7** [comment]
This script may be helpful to prepare `asdf` clean environment quickly:

```sh
# prepare
git clone https://github.com/elixir-ecto/ecto_sql.git
cd ecto_sql

# ensure that there is no other PostgreSQL server started

# install asdf
git clone https://github.com/asdf-vm/asdf.git ~/.asdf --branch v0.7.4

# Linux | Bash
echo -e '\n. $HOME/.asdf/asdf.sh' >> ~/.bashrc
echo -e '\n. $HOME/.asdf/completions/asdf.bash' >> ~/.bashrc
source ~/.bash

# MacOS | Bash
echo -e '\n. $HOME/.asdf/asdf.sh' >> ~/.bash_profile
echo -e '\n. $HOME/.asdf/completions/asdf.bash' >> ~/.bash_profile
source ~/.bash_profile

# plugins
asdf plugin-add elixir
asdf plugin-add erlang
asdf plugin-add postgres

# installation
asdf install elixir 1.9.2-otp-22
asdf install erlang 22.1.5
asdf install postgres 12.0

# change current versions
asdf local elixir 1.9.2-otp-22
asdf local erlang 22.1.5
asdf local postgres 12.0

# prepare PostgreSQL
pg_ctl start
psql -c "create database \"$USER\" with owner \"$USER\" encoding 'UTF8' LC_COLLATE = 'pl_PL.UTF-8' LC_CTYPE = 'pl_PL.UTF-8' TEMPLATE template0;"

# finalize
mix deps.get
mix test
```

**Comment by Eiji7** [comment]
@josevalim I'm not sure if this would be helpful, but here is a `PostgreSQL` log file from my last test task call:

https://gist.github.com/Eiji7/0706e2ee62381a3b16dd4c33dc9f769b

There you can find messages like:
> 2019-11-01 20:21:59.583 CET [19395] ERROR:  prepared statement "ecto_update_posts" does not exist
https://gist.github.com/Eiji7/0706e2ee62381a3b16dd4c33dc9f769b#file-log-sql-L6617

**Comment by wojtekmach** [comment]
Thank you for the report, we occasionally see this on our CI too.

**Comment by Eiji7** [comment]
@wojtekmach Yes, I know as I have mentioned CI too:
> Here is the error which happen locally and on CI:

Looks like I'm the only one who reproduced it locally as nobody else commented anything in this issue … Do you need something specific to debug? Unfortunately I have no idea why it raises and especially why not every time.

**Comment by leandrocp** [comment]
@Eiji7 thanks for your report, I got the same errors here.

You can see that happening at https://github.com/elixir-ecto/ecto/blob/890cc735b885bf629f271e2b631fdc49ba1276e1/integration_test/cases/repo.exs#L1010 when running `mix test --seed 390015`, and here's a more detailed log:

```
[7/29][idle] DEBUG:  00000: StartTransaction(1) name: unnamed; blockState: DEFAULT; state: INPROGRESS, xid/subid/cid: 0/1/0
[7/29][idle] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][idle] STATEMENT:  BEGIN
[7/29][idle] LOG:  00000: statement: BEGIN
[7/29][idle] LOCATION:  exec_simple_query, postgres.c:975
[7/29][idle in transaction] LOG:  00000: statement: SAVEPOINT postgrex_query
[7/29][idle in transaction] LOCATION:  exec_simple_query, postgres.c:975
[7/29][SAVEPOINT] DEBUG:  00000: StartSubTransaction(1) name: unnamed; blockState: INPROGRESS; state: INPROGRESS, xid/subid/cid: 0/1/0
[7/29][SAVEPOINT] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][SAVEPOINT] STATEMENT:  SAVEPOINT postgrex_query
[7/29][SAVEPOINT] DEBUG:  00000: StartSubTransaction(2) name: postgrex_query; blockState: SUBBEGIN; state: INPROGRESS, xid/subid/cid: 0/2/0
[7/29][SAVEPOINT] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][SAVEPOINT] STATEMENT:  SAVEPOINT postgrex_query
[7/29][idle in transaction] DEBUG:  00000: bind <unnamed> to ecto_insert_posts
[7/29][idle in transaction] LOCATION:  exec_bind_message, postgres.c:1534
[7/29][idle in transaction] ERROR:  26000: prepared statement "ecto_insert_posts" does not exist
[7/29][idle in transaction] LOCATION:  FetchPreparedStatement, prepare.c:515
[7/29][idle in transaction] DEBUG:  00000: AbortSubTransaction(1) name: unnamed; blockState: INPROGRESS; state: INPROGRESS, xid/subid/cid: 0/1/0
[7/29][idle in transaction] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][idle in transaction] DEBUG:  00000: AbortSubTransaction(2) name: postgrex_query; blockState: SUBINPROGRESS; state: INPROGRESS, xid/subid/cid: 0/2/0
[7/29][idle in transaction] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][idle in transaction (aborted)] LOG:  00000: statement: ROLLBACK TO SAVEPOINT postgrex_query;RELEASE SAVEPOINT postgrex_query
[7/29][idle in transaction (aborted)] LOCATION:  exec_simple_query, postgres.c:975
[7/29][ROLLBACK] DEBUG:  00000: CleanupSubTransaction(1) name: unnamed; blockState: INPROGRESS; state: INPROGRESS, xid/subid/cid: 0/1/0
[7/29][ROLLBACK] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][ROLLBACK] STATEMENT:  ROLLBACK TO SAVEPOINT postgrex_query;RELEASE SAVEPOINT postgrex_query
[7/29][ROLLBACK] DEBUG:  00000: CleanupSubTransaction(2) name: unnamed; blockState: SUBABORT_RESTART; state: ABORT, xid/subid/cid: 0/2/0
[7/29][ROLLBACK] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][ROLLBACK] STATEMENT:  ROLLBACK TO SAVEPOINT postgrex_query;RELEASE SAVEPOINT postgrex_query
[7/29][ROLLBACK] DEBUG:  00000: StartSubTransaction(1) name: unnamed; blockState: INPROGRESS; state: INPROGRESS, xid/subid/cid: 0/1/0
[7/29][ROLLBACK] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][ROLLBACK] STATEMENT:  ROLLBACK TO SAVEPOINT postgrex_query;RELEASE SAVEPOINT postgrex_query
[7/29][ROLLBACK] DEBUG:  00000: StartSubTransaction(2) name: postgrex_query; blockState: SUBBEGIN; state: INPROGRESS, xid/subid/cid: 0/3/0
[7/29][ROLLBACK] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][ROLLBACK] STATEMENT:  ROLLBACK TO SAVEPOINT postgrex_query;RELEASE SAVEPOINT postgrex_query
[7/29][RELEASE] DEBUG:  00000: CommitSubTransaction(1) name: unnamed; blockState: INPROGRESS; state: INPROGRESS, xid/subid/cid: 0/1/0
[7/29][RELEASE] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][RELEASE] STATEMENT:  ROLLBACK TO SAVEPOINT postgrex_query;RELEASE SAVEPOINT postgrex_query
[7/29][RELEASE] DEBUG:  00000: CommitSubTransaction(2) name: postgrex_query; blockState: SUBRELEASE; state: INPROGRESS, xid/subid/cid: 0/3/0
[7/29][RELEASE] LOCATION:  ShowTransactionStateRec, xact.c:5175
[7/29][RELEASE] STATEMENT:  ROLLBACK TO SAVEPOINT postgrex_query;RELEASE SAVEPOINT postgrex_query
[7/29][idle in transaction] LOG:  00000: statement: ROLLBACK
[7/29][idle in transaction] LOCATION:  exec_simple_query, postgres.c:975
```

`log_line_prefix` is configured to `'[%v][%i] '`

In that log I don't see a `parse ecto_insert_posts` that usually happens for each insert, and also that error goes away if you provide a value on a field for `%Post{}`, like `%Post{title: "whatever"}`.

**Comment by josevalim** [comment]
My hunch is that this is a race condition based on the SQL Sandbox. We could try changing it to `%Post{title: "whatever"}` and see if the issue is addressed permanently or just makes it less likely. WDYT?

**Comment by leandrocp** [comment]
My first attempt was to turn `async` off but it didn't help, also note there's another place this is happening.

Do you mean the Sandbox adapter? If that's the case I'd say to debug it first and make sure it's okay to avoid causing the same error on projects that are using Ecto, in other words I believe that Repo.insert should work, right?

**Comment by josevalim** [comment]
@leandrocp how did you turn async off? Did you try running the suite with the `--trace` flag?

**Comment by leandrocp** [comment]
Btw running with `--trace` I get

```
  * test update all with casting and dumping on id type field (0.6ms)

  1) test update all with casting and dumping on id type field (Ecto.Integration.RepoTest)
     /Users/leandro/code/github/leandrocp/ecto/integration_test/cases/repo.exs:1009
     ** (Postgrex.Error) ERROR 26000 (invalid_sql_statement_name) prepared statement "ecto_insert_posts" does not exist
     code: assert %Post{id: id1} = TestRepo.insert!(%Post{})
     stacktrace:
       (ecto_sql 3.4.4) lib/ecto/adapters/sql.ex:593: Ecto.Adapters.SQL.raise_sql_call_error/1
       (ecto 3.4.4) lib/ecto/repo/schema.ex:661: Ecto.Repo.Schema.apply/4
       (ecto 3.4.4) lib/ecto/repo/schema.ex:263: anonymous fn/15 in Ecto.Repo.Schema.do_insert/4
       (ecto 3.4.4) lib/ecto/repo/schema.ex:164: Ecto.Repo.Schema.insert!/4
       /Users/leandro/code/github/leandrocp/ecto/integration_test/cases/repo.exs:1010: (test)
```

**Comment by josevalim** [comment]
Wait, if you can consistently reproduce it then we should pair on it. Does it even happen if you run the test in isolation?

**Comment by leandrocp** [comment]
I can reproduce it by running ` mix test --seed 390015 ../ecto/integration_test/cases/repo.exs`, I've tried several times and it always fail. But I can't reproduce it by running only that test in particular, it's always green.

**Comment by josevalim** [comment]
Fixed in Postgrex on v0.15.5. I can confirm it only affected the sandbox in the specific scenarios you issued a query that was invalid.

**Comment by leandrocp** [comment]
Ref https://github.com/elixir-ecto/postgrex/commit/71e043cbdf2ca40297845ff0d1c5d66831ed5a3d
